File: translated/f-to-c/DRB149-missingdata1-orig-gpu-no.c
/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Copyright (c) 2017-20, Lawrence Livermore National Security, LLC
and DataRaceBench project contributors. See the DataRaceBench/COPYRIGHT file for details.

SPDX-License-Identifier: (BSD-3-Clause)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/

//Classic i-k-j matrix multiplication. No data race pairs.

#include <omp.h>
#include <stdio.h>
#include <stdlib.h>

int main()
{
    int len, i, j;
    int* a;
    int* b;
    int* c;

    len = 100;

    a = (int*)malloc(len * sizeof(int));
    b = (int*)malloc((len + len * len) * sizeof(int));
    c = (int*)malloc(len * sizeof(int));

    for (i = 0; i < len; i++) {
        for (j = 0; j < len; j++) {
            b[j + i * len] = 1;
        }
        a[i] = 1;
        c[i] = 0;
    }

    #pragma omp target map(to:a[0:len],b[0:len+len*len]) map(tofrom:c[0:len]) device(0)
    #pragma omp teams distribute parallel for
    for (i = 0; i < len; i++) {
        for (j = 0; j < len; j++) {
            c[i] = c[i] + a[j] * b[j + i * len];  // No race: each thread works on different c[i]
        }
    }

    for (i = 0; i < len; i++) {
        if (c[i] != len) {
            printf("%d\n", c[i]);
        }
    }

    free(a);
    free(b);
    free(c);

    return 0;
} 

File: translated/f-to-go/DRB149-missingdata1-orig-gpu-no.go
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// Copyright (c) 2017-20, Lawrence Livermore National Security, LLC
// and DataRaceBench project contributors. See the DataRaceBench/COPYRIGHT file for details.
//
// SPDX-License-Identifier: (BSD-3-Clause)
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

//Classic i-k-j matrix multiplication. No data race pairs.

package main

import (
	"fmt"
	"runtime"
	"sync"
)

func main() {
	var length int
	var a, b, c []int

	length = 100

	a = make([]int, length)
	b = make([]int, length+length*length)
	c = make([]int, length)

	// Initialize arrays
	for i := 1; i <= length; i++ {
		for j := 1; j <= length; j++ {
			b[j-1+(i-1)*length] = 1
		}
		a[i-1] = 1
		c[i-1] = 0
	}

	//$omp target map(to:a,b) map(tofrom:c) device(0)
	//$omp teams distribute parallel do
	var wg sync.WaitGroup
	numTeams := runtime.NumCPU()
	chunkSize := length / numTeams
	if chunkSize < 1 {
		chunkSize = 1
	}

	for team := 0; team < numTeams; team++ {
		start := team*chunkSize + 1
		end := start + chunkSize - 1
		if team == numTeams-1 {
			end = length
		}
		wg.Add(1)
		go func(start, end int) {
			defer wg.Done()
			for i := start; i <= end; i++ {
				for j := 1; j <= length; j++ {
					c[i-1] = c[i-1] + a[j-1]*b[j-1+(i-1)*length] // No race - proper partitioning
				}
			}
		}(start, end)
	}
	wg.Wait()
	//$omp end teams distribute parallel do
	//$omp end target

	for i := 1; i <= length; i++ {
		if c[i-1] != length {
			fmt.Printf("%d\n", c[i-1])
		}
	}

	// deallocate(a,b,c)
}